src/lib/zkstring.ts

```diff
@@ -0,0 +1,85 @@
+import { Bool, Circuit, Field } from '../snarky';
+import { CircuitValue, prop } from './circuit_value';
+
+export class Character extends CircuitValue {
+  @prop value: Field;
+
+  constructor(value: Field) {
+    super();
+
+    Character.validateInput(value);
+    this.value = value;
+  }
+
+  equals(char: Character): Bool {
+    return this.value.equals(char.value);
+  }
+
+  toField(): Field {
+    return this.value;
+  }
+
+  toString(): string {
+    const charCode = Number(this.value.toString());
+    return String.fromCharCode(charCode);
+  }
+
+  static fromString(str: string) {
+    const char = Field(str.charCodeAt(0));
+    return new Character(char);
+  }
+
+  static validateInput(value: Field): Boolean {
+    if (true) {
+      return true;
+    } else {
+      throw new Error('Character does not pass validation')
+    }
+  }
+};
+
+export class ZKString extends CircuitValue {
+  @prop values: Character[];
+
+  constructor(values: Character[]) {
+    super();
+
+    this.values = values;
+  }
+
+  length(): Field {
+    return Field(this.values.length);
+  }
+
+  substring(start: number, end: number): ZKString {
+    return new ZKString(this.values.slice(start, end));
+  }
+
+  toFields(): Field[] {
+    return this.values.map(x => x.toField());
+  }
+
+  toString(): string {
+    return this.values.map(x => x.toString()).join('');
+  }
+
+  equals(str: ZKString): Bool {
+    try {
```

Comment:
Not sure this is valid in a circuit.  I had some index out of bounds issues with some examples, and thought it would be nice to ensure that lengths are equal before continuing.